client/transaction-pool/src/revalidation.rs

```diff
@@ -0,0 +1,346 @@
+		};
+
+		match api.validate_transaction(&BlockId::Number(at), ext.data.clone()).await {
+			Ok(Err(TransactionValidityError::Invalid(err))) => {
+				log::debug!(target: "txpool", "[{:?}]: Removed during revalidation: {:?}", ext_hash, err);
+				invalid_hashes.push(ext_hash);
+			},
+			Ok(Err(TransactionValidityError::Unknown(err))) => {
+				// skipping unknown, they might be pushed by valid or invalid transaction
+				// when latter resubmitted.
+				log::debug!(target: "txpool", "[{:?}]: Unknown during revalidation: {:?}", ext_hash, err);
+			},
+			Ok(Ok(validity)) => {
+				let data = ext.data.clone();
+				let (_, bytes) = api.hash_and_length(&data);
+				revalidated.insert(
+					ext_hash.clone(),
+					ValidatedTransaction::Valid(
+						Transaction {
+							data: ext.data.clone(),
+							bytes,
+							hash: ext_hash.clone(),
+							priority: validity.priority,
+							requires: validity.requires,
+							provides: validity.provides,
+							propagate: validity.propagate,
+							valid_till: at
+								.saturated_into::<u64>()
+								.saturating_add(validity.longevity),
+					}),
+				);
+			},
+			Err(validation_err) => {
+				log::debug!(
+					target: "txpool",
+					"[{:?}]: Error during revalidation: {:?}. Removing.",
+					ext_hash,
+					validation_err
+				);
+				invalid_hashes.push(ext_hash);
+			}
+		}
+	}
+
+	pool.remove_invalid(&invalid_hashes);
+	pool.resubmit(revalidated);
+}
+
+impl<Api: ChainApi> RevalidationWorker<Api> {
+	fn new(
+		api: Arc<Api>,
+		pool: Arc<Pool<Api>>,
+		from_queue: mpsc::UnboundedReceiver<WorkerPayload<Api>>
+	) -> Self {
+		Self {
+			api,
+			pool,
+			from_queue,
+			tick_timeout: Box::pin(interval(BACKGROUND_REVALIDATION_INTERVAL)),
+			block_ordered: Default::default(),
+			members: Default::default(),
+			best_block: Zero::zero(),
+		}
+	}
+
+	fn prepare_batch(&mut self) -> Vec<ExHash<Api>> {
+		let mut queued_exts = Vec::new();
+		let mut empty = Vec::<NumberFor<Api>>::new();
+
+		// Take maximum of count transaction by order
+		// which they got into the pool
+		for (block_number, ext_map) in self.block_ordered.iter_mut() {
+			if queued_exts.len() >= BACKGROUND_REVALIDATION_BATCH_SIZE { break; }
+
+			loop {
+				let next_key = match ext_map.iter().nth(0) {
+					Some(k) => k.clone(),
+					None => { break; }
+				};
+
+				ext_map.remove(&next_key);
+				self.members.remove(&next_key);
+
+				queued_exts.push(next_key);
+
+				if ext_map.len() == 0 { empty.push(*block_number); }
+
+				if queued_exts.len() >= BACKGROUND_REVALIDATION_BATCH_SIZE { break; }
+			}
+		}
+
+		// retain only non-empty
+		for empty_block_number in empty.into_iter() { self.block_ordered.remove(&empty_block_number); }
+
+		queued_exts
+	}
+
+	fn push(&mut self, worker_payload: WorkerPayload<Api>) {
+		// we don't add something that already scheduled for revalidation
+		let transactions = worker_payload.transactions;
+		let block_number = worker_payload.at;
+
+		for ext_hash in transactions.into_iter() {
```

Comment:
```suggestion
                let transactions = transactions.into_iter().filter(|hash| !self.members.contains_key(&hash));
		for ext_hash in transactions {
		    ...
		}
```
Or just remove `into_iter` from this line if you prefer to keep `contains_key` within the loop.